enterprise/backend/src/metabase_enterprise/data_editing/api.clj

```diff
@@ -214,6 +242,58 @@
     {:table_id table-id
      :token token}))
 
+(defn translate-undo-error [e]
+  (case (:error (ex-data e))
+    :undo/none            (ex-info (tru "Nothing to do")                                         {:status-code 204} e)
+    :undo/cannot-undelete (ex-info (tru "You cannot undo your previous change.")                 {:status-code 405} e)
+    :undo/conflict        (ex-info (tru "Your previous change has a conflict with another edit") {:status-code 409} e)
+    e))
+
+(api.macros/defendpoint :post "/undo"
+  "Undo the last change you made.
+  For now only supports tables, but in future will support editables for sure.
+  Maybe actions, workflows, etc.
+  Could even generalize to things like edits to dashboard definitions themselves."
+  [_
+   _
+   {:keys [table-id no-op]}] :- [:map
+                                 [:table-id ms/PositiveInt]
+                                 [:no-op {:optional true} ms/BooleanValue]]
+  (check-permissions)
+  (api/check-404 (t2/select-one-pk :model/Table table-id))
+  (let [user-id api/*current-user-id*]
+    (if no-op
+      {:batch_num (undo/next-batch-num true user-id table-id)}
+      ;; IDEA encapsulate this in an action
+      ;; IDEA use generic action calling API instead of having this endpoint
+      (try
+        {:result (undo/undo! user-id table-id)}
+        (catch ExceptionInfo e
+          (throw (translate-undo-error e)))))))
+
+(api.macros/defendpoint :post "/redo"
```

Comment:
Feels like a lot of boilerplate having both endpoints, maybe redo should be a flag?